[DJM-1220] Add Private Link Environments (Preview) page for Databricks (#36177)
DataDog/documentation: content/en/data_observability/jobs_monitoring/databricks/private_link.md

The request flow works as follows:

1. The [Private Action Runner][2] reaches out to the Datadog backend through the [VPC Endpoint][3] or over the public internet to query for pending requests. If a request is found, details are returned to the runner.
1. Databricks credentials (client ID and secret) are retrieved from secret storage. A token is generated for the session by calling the Databricks API from the runner using those credentials.
1. The runner executes the fetched query against the Databricks API. The results are returned to the runner.
1. The runner forwards the results back to the Datadog backend for processing.

All authentication to Datadog is handled by the underlying Private Action Runner. A locally configured allowlist is set on the runner during installation to restrict the actions that can be performed to only those that are necessary. You can also configure access controls within Datadog to manage which users and teams can use the runner. See [Manage access to Private Action Runners][16] for more information.

## Installation

> ### Step 1: Datadog prerequisites

Author: Something we learned just today - Remote Config is required to be enabled for Private Action Runners to function properly. Could we add a step under the prerequisites to that effect after service account creation? 

```
1. Ensure [Remote Config](https://app.datadoghq.com/organization-settings/remote-config) is enabled for your Datadog account
```

     - **AWS**: [Set up VPC Endpoints][3] (available in US1, AP1, and AP2)
     - **Azure**: [Private Link service setup][6] (available in US3 only). This requires approval in the Datadog Azure account. Reach out to Datadog Support as specified in the instructions.
     - **GCP**: [Set up Private Service Connect Endpoints][7] (available in US5 and EU1)
1. Create a Datadog service account:
   1. Go to the [Service Accounts][8] page.
   1. Click **New Service Account**.
   1. Enter a name and email address for your service account.
   1. Under **Assign Roles**, select the **Datadog Standard Role**. Alternatively, select a custom role that has the `Connections Resolve` permission.
   1. Click **Create Service Account**.

### Step 2: Databricks prerequisites

1. Create a Databricks service principal ([Databricks documentation][9]):
   1. As a workspace admin, log in to the Databricks workspace.
   1. Click your username in the top bar and select **Settings**.
   1. Click the **Identity and access** tab.
   1. Next to **Service principals**, click **Manage**.
   1. Click **Add service principal**, then click **Add new** and enter a name.

>       - Alternatively, to reuse a service principal from another workspace, select an existing service principal and skip the next step.

Author: A note here that we don't support Entra managed IDs might be useful here, included a screenshot of the modal for our own reference (we don't have to include it in the docs)

<img width="687" height="477" alt="Image" src="https://github.com/user-attachments/assets/0ec1d824-db49-4460-bd39-3ee2a1c6e5a6" />

```
- We currently do not support Microsoft Entra ID managed Service Principals over Private Link at this time.
```

1. Add the service principal as a workspace admin. Alternatively, assign more granular permissions following the instructions in [Permissions][11].
   1. Click the **Identity and access** tab.
   1. Next to **Groups**, click **Manage**.
   1. Select the **admins** group.
   1. Click **Add members** and select the service principal added above.
   1. Click **Add**.

### Step 3: Set up the Private Action Runner

Set up your Private Action Runner using **one** of the following options.

{{< tabs >}}

{{% tab "Helm (Kubernetes)" %}}

1. Install the Private Action Runner into your Kubernetes cluster ([additional docs][1], [Helm chart][2]):
   1. Go to the [Private Action Runner setup][3] page.
   1. Select the **Script** option, then click **Next**.
   1. Follow the instructions under **Kubernetes**.
   1. The Private Action Runner should show up at the bottom as "successfully installed."
1. Update the existing `values.yaml` (referencing the [Helm chart values][4]):
   1. Set `image.repository` to `gcr.io/datadoghq/dd-data-observability-par`.
   1. Set `image.tag` to `1.21.0-0`.

> 1. If using cloud secret storage: assign an identity ([Workload Identity][5] or [IAM Role][6]) to the pod with permissions to read the secret created in [Step 2](#step-2-databricks-prerequisites).

Author: Is it worth reminding the user here (and in the Docker tab) of the env var options that are available to them? e.g.:

```
1. Configure credentials on the Private Action Runner using one of the following methods:
    1. Set the `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` environment variables directly on the runner. Leave "Secret Path" in the integration tile blank.
    2. Use cloud secret storage. Ensure an identity is assigned to the pod ([Workload Identity][5] or [IAM Role][6]) with permissions to read the secret created in [Step 2](#step-2-databrick-prerequisites), and provide the path to the secret either via the `DATABRICKS_SECRET_PATH` environment variable, or by providing it to the "Secret Path" field in the integration tile.
```

1. Set up the Databricks integration:
   1. Go to the [Databricks integration][13] tile.
   1. Under **Credentials**, select **Private Action Runner**.
      1. Under **Connection**, select the connection created in the previous step.
      1. Under **Service Account**, select the service account created in [Step 1](#step-1-datadog-prerequisites).
      1. Under **Secret Path**, enter the secret path from [Step 2](#step-2-databricks-prerequisites). Leave this field blank if you set credentials using the `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` environment variables, or if you set the path using the `DATABRICKS_SECRET_PATH` environment variable on the runner.
   1. Complete the remainder of the integration setup as described in [Configure the Datadog-Databricks integration][14]:
      1. Enter the **Workspace Name**, **Workspace URL**, and **System Tables SQL Warehouse ID**.
      1. Enable the products you are interested in.
         **Note:** Model Serving behind Private Link is not supported at this time.
   1. Click **Save Databricks Workspace**.
   1. Verify there are no errors on save.

## Validation

> After completing the setup, jobs and schema information should appear in [Data Observability: Jobs Monitoring][15] shortly.

Author: Schema info comes in via a different product, "jobs and cluster information" might be more suitable here.

---

> title: "Private Link Environments"

Reviewer: ```suggestion
title: "Private Link Connectivity"
```

> description: "Set up Data Observability: Jobs Monitoring for Databricks workspaces deployed in Private Link environments using a Private Action Runner."

Reviewer: ```suggestion
description: "Set up Data Observability: Jobs Monitoring for Databricks workspaces deployed with Private Link Connectivity using a Private Action Runner."
```

further_reading:
    - link: '/data_observability/jobs_monitoring/databricks'
      tag: 'Documentation'
      text: 'Data Observability: Jobs Monitoring for Databricks'
    - link: '/actions/private_actions'
      tag: 'Documentation'
      text: 'Private Action Runners'
---

> <div class="alert alert-info">Monitoring Databricks in Private Link environments is in Preview.</div>

Reviewer: ```suggestion
<div class="alert alert-info">Monitoring Databricks via Private Link Connectivity is in Preview.</div>
```

> Databricks workspaces deployed in [Private Link environments][1] are isolated from the public internet, which prevents Datadog crawlers from accessing the Databricks API directly. To monitor these workspaces, deploy a [Private Action Runner][2] (PAR) in your environment. The runner polls Datadog for requests, executes them against the Databricks API locally, and sends the results back through your existing [Datadog VPC Endpoint][3].

Reviewer: ```suggestion
Databricks workspaces deployed using [Private Link Connectivity][1] are isolated from the public internet, which prevents Datadog from accessing the Databricks APIs directly. To monitor these workspaces, deploy a [Private Action Runner][2] (PAR) in your environment. The runner polls Datadog for requests, executes them against the Databricks API locally, and sends the results back through your existing [Datadog VPC Endpoint][3].
```

      <div class="alert alert-info">Datadog does not support Microsoft Entra ID managed Service Principals over Private Link.</div>
   1. Click **Add**.
1. Generate secrets for the Databricks service principal ([Databricks documentation][10]):
   1. Select the service principal created above.
   1. Click the **Secrets** tab, then click **Generate secret**.
   1. Set the secret's lifetime in days (maximum 730 days).
   1. Click **Generate**.
   1. Make the Databricks credentials available to the runner using **one** of the following methods:
      - **Cloud secret storage** (AWS Secrets Manager, AWS Parameter Store, or Azure Key Vault): Store the credentials as a JSON blob:
        ```json
        {"client_id": "<CLIENT_ID>", "client_secret": "<CLIENT_SECRET>"}
        ```
        Make note of the secret path (AWS ARN or AKV URL). You enter this in the integration tile or set it as the `DATABRICKS_SECRET_PATH` environment variable on the runner container or pod.
      - **Environment variables**: Set `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` directly on the runner container or pod. When using this method, leave the **Secret Path** field blank in the integration tile.

> 1. Add the service principal as a workspace admin. Alternatively, assign more granular permissions following the instructions in [Permissions][11].

Reviewer: Do we really need workspace admin? i.e. are we installing the init script globally via this install path?

In addition to linking to the granular permissions if users want to see those details, should we also just call out the servive principal needs Workspace access and Databricks SQL access entitlements?

Author: Yes, workspace admin is still ideal for us if possible - we do attempt to install the global init script. Good point on the workspace & Databricks SQL access, will link to the other page with that info

Author: Tried to add this as best I could, with System Table as a separate step. Let me know what you think!

   1. Select **Create and Test**. Verify the test results do not return an error.
1. Set up the Databricks integration:
   1. Go to the [Databricks integration][13] tile.
   1. Under **Credentials**, select **Private Action Runner**.
      1. Under **Connection**, select the connection created in the previous step.
      1. Under **Service Account**, select the service account created in [Step 1](#step-1-datadog-prerequisites).
      1. Under **Secret Path**, enter the secret path from [Step 2](#step-2-databricks-prerequisites). Leave this field blank if you set credentials using the `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` environment variables, or if you set the path using the `DATABRICKS_SECRET_PATH` environment variable on the runner.
   1. Complete the remainder of the integration setup as described in [Configure the Datadog-Databricks integration][14]:
      1. Enter the **Workspace Name**, **Workspace URL**, and **System Tables SQL Warehouse ID**.
      1. Enable the products you are interested in.

>          **Note:** Model Serving behind Private Link is not supported at this time.

Reviewer: Add callout + docs link to deploy the init script for monitoring jobs & clusters on classic compute. Also step for restarting clusters.

        ```json
        {"client_id": "<CLIENT_ID>", "client_secret": "<CLIENT_SECRET>"}
        ```
        Make note of the secret path (AWS ARN or AKV URL). You enter this in the integration tile or set it as the `DATABRICKS_SECRET_PATH` environment variable on the runner container or pod.
      - **Environment variables**: Set `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` directly on the runner container or pod. When using this method, leave the **Secret Path** field blank in the integration tile.
1. Provision workspace access and entitlements for the service principal. See [Permissions][11] for full details.
   1. Grant the service principal **Workspace Admin** privileges. This allows Datadog to manage init script installations and updates automatically. Alternatively, assign more [granular permissions][11] for monitoring jobs, clusters, and queries.
      1. Click the **Identity and access** tab.
      1. Next to **Groups**, click **Manage**.
      1. Select the **admins** group.
      1. Click **Add members** and select the service principal added above.
      1. Click **Add**.

>    1. To access Databricks cost data, grant the service principal `CAN USE` on the SQL Warehouse and read access to [system tables][20] in Unity Catalog. See [Permissions][11] for details.

Reviewer: ```suggestion
   1. To access Databricks cost data or monitor serverless jobs, grant the service principal `CAN USE` on the SQL Warehouse and read access to [system tables][20] in Unity Catalog. See [Permissions][11] for details.
```
